### DISCUSSION TITLE
why yolov8 uses torch.zeros(1, ch, s, s) as input to forward pass to calculate model stride ? can someone explain me will different task will have different stride

### DISCUSSION THREAD

**Discussion by rakeshKM** [ultralytics/ultralytics#4492] [Q&A]
reference  task.py file:

# Build strides
m = self.model[-1]  # Detect()
if isinstance(m, (Detect, Segment, Pose)):
    s = 256  # 2x min stride
    m.inplace = self.inplace
    forward = lambda x: self.forward(x)[0] if isinstance(m, (Segment, Pose)) else self.forward(x)
    m.stride = torch.tensor([s / x.shape[-2] for x in forward(torch.zeros(1, ch, s, s))])  # forward
    self.stride = m.stride
    m.bias_init()  # only run once
else:
    self.stride = torch.Tensor([32])  # default stride for i.e. RTDETR

**Comment by crackwitz** [comment]
related: https://stackoverflow.com/questions/76969821/why-yolov8-uses-torch-zeros1-ch-s-s-as-input-to-forward-pass-to-calculate-m

**Reply by glenn-jocher** [reply]
@crackwitz in YOLOv8, the stride calculation using `torch.zeros(1, ch, s, s)` as input to the forward pass is a method to determine the stride of the feature maps relative to the input image. The stride is the step size with which the model's convolutional filters move across the image. It's crucial for scaling the output feature maps back to the original image dimensions during tasks like detection, segmentation, and pose estimation.

The reason for using a tensor of zeros is that we are not interested in the actual output values during this stride calculation; we only need the dimensions of the output feature maps to compute the stride. By passing a tensor with a known size through the model, we can observe how the dimensions change and calculate the stride accordingly.

Different tasks may indeed have different strides because the architecture of the model's layers and how they process the input can vary depending on the task. For example, segmentation models might have a different series of layers that preserve spatial resolution more than detection models, leading to different stride values.

The stride is then stored in `m.stride`, which is used later during inference to correctly scale and interpret the model's predictions relative to the input image size. This step is essential for accurate localization of objects, segmentation masks, or keypoints in the original image space. The stride calculation is a one-time setup operation that configures the model for subsequent forward passes with actual data.

**Reply by crackwitz** [reply]
Not my issue though

**Reply by glenn-jocher** [reply]
Nice, thanks for the stackoverflow link!

**Reply by minhhb9x** [reply]
The coord box edge is encoded in DFL form with `reg_max=16` if I'm right. However, I still don't know whether the label box is scaled by stride to 3 different heads or not to calculate the loss function. Can you explain this, please?

**Reply by glenn-jocher** [reply]
The label box is indeed scaled by the stride for each detection head to calculate the loss function accurately. This ensures alignment between the feature map and the original image dimensions.

**Reply by minhhb9x** [reply]
Assume that I have a label box `x_center, y_center, w, h` in relative coordinates. I read that `reg_max=16` means there are 16 bins for the probability offset of an anchor. How model set the label for DFL layer of a head to calculate DFLoss? Can you explain more detail, please? Thank you very much.

**Reply by glenn-jocher** [reply]
The label for the DFL layer is set by discretizing the continuous offset into `reg_max=16` bins, representing the probability distribution of the offset from the anchor. This helps the model learn precise localization by predicting the most likely bin for each coordinate.
